stingray/fourier.py

```diff
@@ -0,0 +1,1286 @@
+    P2noise : float
+        Poisson noise level of the reference-band periodogram
+    Nave : int
+        number of intervals that have been averaged to obtain the input spectra
+
+    Other Parameters
+    ----------------
+    intrinsic_coherence : float, default 1
+        If known, the intrinsic coherence.
+    """
+    if Nave > 500:
+        return 0. * P1
+    bsq = P1 * P2 - intrinsic_coherence * (P1 - P1noise) * (P2 - P2noise)
+    return bsq / Nave
+
+
+def raw_coherence(C, P1, P2, P1noise, P2noise, Nave, intrinsic_coherence=1):
+    """Raw coherence from Ingram 2019.
+
+    Parameters
+    ----------
+    C : complex `np.array`
+        cross spectrum
+    P1 : float `np.array`
+        sub-band periodogram
+    P2 : float `np.array`
+        reference-band periodogram
+    P1noise : float
+        Poisson noise level of the sub-band periodogram
+    P2noise : float
+        Poisson noise level of the reference-band periodogram
+    Nave : int
+        number of intervals that have been averaged to obtain the input spectra
+
+    Other Parameters
+    ----------------
+    intrinsic_coherence : float, default 1
+        If known, the intrinsic coherence.
+    """
+    bsq = bias_term(C, P1, P2, P1noise, P2noise, Nave, intrinsic_coherence=intrinsic_coherence)
+    num = (C * np.conj(C)).real - bsq
+    if isinstance(num, Iterable):
+        num[num < 0] = (C * np.conj(C)).real[num < 0]
+    elif num < 0:
+        num = (C * np.conj(C)).real
+    den = P1 * P2
+    return num / den
+
+
+def estimate_intrinsic_coherence(C, P1, P2, P1noise, P2noise, Nave):
+    """Estimate intrinsic coherence
+
+    Use the iterative procedure from sec. 5 of Ingram 2019
+
+    Parameters
+    ----------
+    C : complex `np.array`
+        cross spectrum
+    P1 : float `np.array`
+        sub-band periodogram
+    P2 : float `np.array`
+        reference-band periodogram
+    P1noise : float
+        Poisson noise level of the sub-band periodogram
+    P2noise : float
+        Poisson noise level of the reference-band periodogram
+    Nave : int
+        number of intervals that have been averaged to obtain the input spectra
+
+    """
+    new_coherence = np.ones_like(P1)
+    old_coherence = np.zeros_like(P1)
+    count = 0
+    while not np.allclose(new_coherence, old_coherence, atol=0.01) and count < 40:
+        # TODO: make it only iterate over the places at low coherence
+        old_coherence = new_coherence
+        bsq = bias_term(C, P1, P2, P1noise, P2noise, Nave, intrinsic_coherence=new_coherence)
+        # old_coherence = new_coherence
+        den = (P1 - P1noise) * (P2 - P2noise)
+        num = (C * C.conj()).real - bsq
+        num[num < 0] = (C * C.conj()).real[num < 0]
+        new_coherence = num / den
+        count += 1
+
+    return new_coherence
+
+
+def error_on_averaged_cross_spectrum(C, Ps, Pr, Nave, Psnoise, Prnoise, common_ref="False"):
+    """Error on cross spectral quantities, From Ingram 2019.
```

Comment:
This should probably come with a warning that this is only valid above 50 averaged spectra or so, if I remember correctly?